src/function/numeric/solveODE.js

```diff
@@ -0,0 +1,254 @@
+  {
+    typed,
+    add,
+    subtract,
+    multiply,
+    divide,
+    max,
+    map,
+    abs,
+    isPositive,
+    isNegative,
+    larger,
+    smaller
+  }
+) => {
+  /**
+     * Numerical Integration of Ordinary Differential Equations
+     *
+     * Two variable step methods are provided:
+     * - "RK23": Bogacki–Shampine method
+     * - "RK45": Dormand-Prince method RK5(4)7M (default)
+     *
+     * The arguments are expected as follows.
+     *
+     * - `func` should be the forcing function `f(t,y)`
+     * - `tspan` should be a vector of two numbers or units `[tStart, tEnd]`
+     * - `y0` the initial values, should be a scalar or a flat array
+     * - `options` should be an object with the following information:
+     *   - `method` ('RK45'): ['RK23', 'RK45']
+     *   - `tol` (1e-3): A numeric value
+     *   - `firstStep`: A numeric value or unit
+     *   - `minStep`: minimum step size of the method
+     *   - `maxStep`: maximum step size of the method
+     *   - `minDelta` (0.2): minimum ratio of change for the step
+     *   - `maxDelta` (5): maximum ratio of change for the step
+     *   - `maxIter` (1e4): maximum number of iterations
+     *
+     * Syntax:
+     *
+     *     math.solveODE(func, tspan, y0)
+     *     math.solveODE(func, tspan, y0, options)
+     *
+     * Examples:
+     *
+     *     function func(t, y) {return y}
+     *     const tspan = [0, 4]
+     *     const y0 = 1
+     *     math.solveODE(func, tspan, y0)
+     *     math.solveODE(func, tspan, [1, 2])
+     *     math.solveODE(func, tspan, y0, { method:"RK23", maxStep:0.1 })
+     *
+     * See also:
+     *
+     *     simplifyCore, derivative, evaluate, parse, rationalize, resolve
+     *
+     * @param {function} func The forcing function f(t,y)
+     * @param {Array | Matrix} tspan The time span
+     * @param {number | BigNumber | Unit | Array | Matrix} y0 The initial value
+     * @param {Object} [options] Optional configuration options
+     * @return {Object} Return an object with t and y values as arrays
+     */
+
+  function _rk (butcherTableau) {
+    // generates an adaptive runge kutta method from it's butcher tableau
+
+    return function (f, T, y0, options) {
+      // adaptive runge kutta methods
+
+      const t0 = T[0] // initial time
+      const tf = T[1] // final time
+      const steps = 1 // divide the time in this number of steps
+      const tol = options.tol ? options.tol : 1e-4 // define a tolerance (must be an option)
+      const maxStep = options.maxStep
+      const minStep = options.minStep
+      const minDelta = options.minDelta ? options.minDelta : 0.2
+      const maxDelta = options.maxDelta ? options.maxDelta : 5
+      const maxIter = options.maxIter ? options.maxIter : 10_000 // stop inifite evaluation if something goes wrong
+      const [a, c, b, bp] = [butcherTableau.a, butcherTableau.c, butcherTableau.b, butcherTableau.bp]
+
+      let h = options.firstStep
```

Comment:
I think this should be `let h = options.firstStep !== undefined`, else it doesn't work when firstStep is `0` because that is falsy.

Reply:
In these methods the step should never be 0, so I think it helps that it's falsy. But I don't know if it should throw a warning or an error.

Reply:
Ahh, it is the first step _size_. That will be larger than zero indeed.

Reply:
Can you add a validation for `firstStep` and throw an error when it is zero or negative? I think that is quite relevant, else you get an infinite loop, right?

Reply:
I agree that there are some missing validations, please review the current state:

## When `firstStep` is 0:
- It's resolved as falsy, and the method chooses it for the user without warning.
  - The method also chooses it for the user if `firstStep` is too big, thus not within `tol`.
- If it didn't resolve as falsy, indeed it would result in an infinite loop due to never being able to increase it by multiplying by `delta` until reaching `maxIter`.
- Similar to what happens with `range(0, 3, 0)` that yields `[]` instead of an error.
- As a reference, both **scipy's** `integrate.solve_ivp` and **Octave's** `ode45` throw an error when `first_step` and `InitialStep` are 0.

## When `firstStep` is negative:
- When `tspan` goes backwards, `firstStep` **should be negative**.
  - Similar to what happens with `range(0, -3, -1)` that yields `[0, -1, -2, -3]`.
  - As a reference, **scipy's** `integrate.solve_ivp` can go backwards, but `first_step` can only be positive; otherwise, it throws an error.
    - Thus, `first_step` only works when going forward. In my opinion, it's an error in their implementation.
  - As a reference, **Octave's** `ode45` can go backwards, but `InitialStep` can only be positive; otherwise, it throws an error.
    - Thus, `InitialStep` only works when solving forward. In my opinion, it's an error in their implementation.
- When `tspan` goes forwards but `firstStep` is negative, it returns only the initial conditions `{t: [t0], y: [y0]}`.
  - Similar to what happens with `range(0, 3, -1)` that yields `[]` instead of an error.

## Currently

As it stands right now, I think it's consistent with functions like `range` in the sense that it prioritizes giving a solution rather than throwing an error.

But being more consistent with other implementations of these methods it could:

- Throw an error when `firstStep` is 0, similar to `scipy` and `Octave`
- Throw an error when `tspan` is forwards and `firstStep` is negative, or vice versa (being dissimilar to `range` but better than `scipy` and `Octave`)

I don't know if it is excessive but it also could:
- Throw a specific error when the units are not consistent in `tspan`, `firstStep`, `maxStep` and `minStep`

Reply:
While writing the previous comment I found an error in my interpretation and edited it.

I will try to go with only positive values for `firstStep` and report back.

Reply:
In the end by allowing only positive number for `firstStep`, `minStep` and `maxStep` the logic was simplified a bit by reducing some conditionals.

I included some additional tests for all time dependent variables `tspan[0]`, `tspan[1]`, `firstStep`, `minStep` and `maxStep`

I validated it's mostly consistent with the other references I mentioned.

Reply:
👌  this sounds good!